File: sdk/typescript/src/langchain.ts
import { LatchError } from "./errors.js";
import type { OpenAIDecisionClient, OpenAIDecideOptions } from "./openai.js";

export interface LangChainToolCall {
  id?: string;
  name: string;
  args: Record<string, unknown>;
  type?: string;
}

export interface ProtectedLangChainToolCall extends LangChainToolCall {
  id: string;
}

export interface LangChainToolGuardOptions {
  maxCalls?: number;
  maxArgumentBytes?: number;
  replayCapacity?: number;
}

export interface LangChainToolCallRequest {
  toolCall: LangChainToolCall;
  tool?: unknown;
  runtime?: { signal?: AbortSignal };
}

export type LangChainToolCallHandler = (request: any) => any;
export type LangChainToolCallHook = (
  request: LangChainToolCallRequest,
  handler: LangChainToolCallHandler,
) => Promise<any>;

export interface LangChainMiddlewareLike {
  readonly name: string;
  readonly wrapToolCall?: LangChainToolCallHook;
  [key: string]: unknown;
}

export interface LangGraphToolNodeLike {
  invoke(input: unknown, config?: unknown): Promise<unknown>;
  [key: string]: unknown;
}

export class LangChainAdapterError extends LatchError {
  override readonly name: string = "LangChainAdapterError";
}

export class LangChainDependencyError extends LangChainAdapterError {
  override readonly name = "LangChainDependencyError";
}

export class LangChainProtocolError extends LangChainAdapterError {
  override readonly name = "LangChainProtocolError";
}

export class LangChainReplayError extends LangChainAdapterError {
  override readonly name = "LangChainReplayError";
  constructor(readonly callId: string) {
    super(`LangChain tool call ${JSON.stringify(callId)} was duplicated or replayed; execution stopped`);
  }
}

export class LangChainNotAllowedError extends LangChainAdapterError {
  override readonly name = "LangChainNotAllowedError";
  constructor(
    readonly callId: string,
    readonly tool: string,
    readonly decision: { readonly decision: string; readonly allowed: boolean },
  ) {
    super(`Latch decision for LangChain tool ${JSON.stringify(tool)} call ${JSON.stringify(callId)} is ${decision.decision}; execution stopped`);
  }
}

export class UnknownLangChainToolError extends LangChainAdapterError {
  override readonly name = "UnknownLangChainToolError";
  constructor(readonly callId: string, readonly tool: string) {
    super(`LangChain tool ${JSON.stringify(tool)} for call ${JSON.stringify(callId)} is not registered; batch not executed`);
  }
}

/** Shared guard for LangChain middleware and full-batch LangGraph execution. */
export class LangChainToolGuard {
  readonly #client: OpenAIDecisionClient;
  readonly #maxCalls: number;
  readonly #maxArgumentBytes: number;
  readonly #replayCapacity: number;
  readonly #replayed = new Set<string>();
  readonly #replayOrder: string[] = [];

  constructor(client: OpenAIDecisionClient, options: LangChainToolGuardOptions = {}) {
    if (client === null || typeof client !== "object" || typeof client.decide !== "function") {
      throw new TypeError("client must provide decide(action)");
    }
    this.#maxCalls = boundedInteger(options.maxCalls ?? 128, 1, 1024, "maxCalls");
    this.#maxArgumentBytes = boundedInteger(options.maxArgumentBytes ?? 1 << 20, 1, 16 << 20, "maxArgumentBytes");
    this.#replayCapacity = boundedInteger(options.replayCapacity ?? 10_000, 128, 1_000_000, "replayCapacity");
    if (this.#replayCapacity < this.#maxCalls) throw new TypeError("replayCapacity cannot be smaller than maxCalls");
    this.#client = client;
  }

  async protect(
    rawCalls: readonly unknown[],
    surface: "langchain_agent" | "langgraph_tool_node",
    options: OpenAIDecideOptions & { allowedTools?: ReadonlySet<string> } = {},
  ): Promise<readonly ProtectedLangChainToolCall[]> {
    if (!Array.isArray(rawCalls)) throw new LangChainProtocolError("tool calls must be an array");
    if (rawCalls.length > this.#maxCalls) throw new LangChainProtocolError(`tool call batch exceeds ${this.#maxCalls} calls`);
    const calls = rawCalls.map((call) => normalizeCall(call, this.#maxArgumentBytes));
    const seen = new Set<string>();
    for (const call of calls) {
      if (seen.has(call.id)) throw new LangChainReplayError(call.id);
      seen.add(call.id);
      if (options.allowedTools !== undefined && !options.allowedTools.has(call.name)) {
        throw new UnknownLangChainToolError(call.id, call.name);
      }
    }
    for (const call of calls) {
      const decision = await this.#client.decide({
        tool: call.name,
        arguments: call.args,
        metadata: {
          protocol: "langchain",
          surface,
          tool_call_id: call.id,
          tool_call_kind: "function",
        },
      }, { signal: options.signal });
      if (!decision.allowed) throw new LangChainNotAllowedError(call.id, call.name, decision);
    }
    this.#reserve(calls);
    return calls;
  }

  #reserve(calls: readonly ProtectedLangChainToolCall[]): void {
    for (const call of calls) if (this.#replayed.has(call.id)) throw new LangChainReplayError(call.id);
    for (const call of calls) {
      this.#replayed.add(call.id);
      this.#replayOrder.push(call.id);
    }
    while (this.#replayOrder.length > this.#replayCapacity) {
      this.#replayed.delete(this.#replayOrder.shift()!);
    }
  }
}

/** Create the native wrapToolCall hook for an existing createMiddleware call. */
export function createLatchToolCallHook(
  client: OpenAIDecisionClient,
  options: LangChainToolGuardOptions = {},
): LangChainToolCallHook {
  const guard = new LangChainToolGuard(client, options);
  return async (request, handler) => {
    if (request === null || typeof request !== "object" || request.tool == null) {
      throw new LangChainProtocolError("tool call has no resolved LangChain tool");
    }
    await guard.protect([request.toolCall], "langchain_agent", { signal: request.runtime?.signal });
    return await handler(request);
  };
}

/** Load LangChain lazily and return middleware ready for createAgent. */
export async function createLatchAgentMiddleware(
  client: OpenAIDecisionClient,
  options: LangChainToolGuardOptions = {},
): Promise<LangChainMiddlewareLike> {
  const langchain = await optionalImport("langchain", 'Install the optional peer dependency with: npm install langchain');
  const createMiddleware = langchain.createMiddleware;
  if (typeof createMiddleware !== "function") throw new LangChainDependencyError("installed langchain does not export createMiddleware");
  return createMiddleware({
    name: "LatchSecurityMiddleware",
    wrapToolCall: createLatchToolCallHook(client, options),
  }) as LangChainMiddlewareLike;
}

/**
 * Build a real LangGraph ToolNode whose invoke boundary preflights every tool
 * call before LangGraph begins its parallel execution.
 */
export async function createLatchToolNode(
  tools: readonly unknown[],
  client: OpenAIDecisionClient,
  nodeOptions: Readonly<Record<string, unknown>> = {},
  guardOptions: LangChainToolGuardOptions = {},
): Promise<any> {
  const prebuilt = await optionalImport(
    "@langchain/langgraph/prebuilt",
    "Install the optional peer dependency with: npm install @langchain/langgraph",
  );
  const ToolNode = prebuilt.ToolNode;
  if (typeof ToolNode !== "function") throw new LangChainDependencyError("installed @langchain/langgraph does not export ToolNode");
  const allowedTools = new Set(tools.map(toolName));
  const guard = new LangChainToolGuard(client, guardOptions);
  const messagesKey = typeof nodeOptions.messagesKey === "string" ? nodeOptions.messagesKey : "messages";

  class ProtectedToolNode extends (ToolNode as new (tools_: readonly unknown[], options_: Readonly<Record<string, unknown>>) => LangGraphToolNodeLike) {
    constructor() {
      super(tools, nodeOptions);
    }

    override async invoke(input: unknown, config?: unknown): Promise<unknown> {
      const calls = extractToolCalls(input, messagesKey);
      const signal = recordOrUndefined(config)?.signal;
      await guard.protect(calls, "langgraph_tool_node", {
        allowedTools,
        signal: signal instanceof AbortSignal ? signal : undefined,
      });
      return await super.invoke(input, config);
    }

    batch(): never {
      throw new LangChainProtocolError("protected ToolNode batch() is disabled; invoke the node through LangGraph");
    }

    stream(): never {
      throw new LangChainProtocolError("protected ToolNode stream() is disabled; stream the compiled LangGraph instead");
    }
  }

  return new ProtectedToolNode();
}

function extractToolCalls(input: unknown, messagesKey: string): readonly unknown[] {
  if (Array.isArray(input)) {
    if (input.length === 0) return [];
    if (input.every(isDirectToolCall)) return input;
    return callsFromMessage(input.at(-1));
  }
  const state = record(input, "LangGraph state");
  const messages = state[messagesKey];
  if (!Array.isArray(messages) || messages.length === 0) {
    throw new LangChainProtocolError(`LangGraph state must contain a non-empty ${JSON.stringify(messagesKey)} array`);
  }
  return callsFromMessage(messages.at(-1));
}

function callsFromMessage(messageValue: unknown): readonly unknown[] {
  const message = record(messageValue, "last LangGraph message");
  const calls = message.tool_calls ?? message.toolCalls;
  if (calls === undefined) return [];
  if (!Array.isArray(calls)) throw new LangChainProtocolError("the last message tool calls must be an array");
  return calls;
}

function isDirectToolCall(value: unknown): boolean {
  if (value === null || typeof value !== "object" || Array.isArray(value)) return false;
  const call = value as Record<string, unknown>;
  return "id" in call && "name" in call && "args" in call;
}

function normalizeCall(value: unknown, maxArgumentBytes: number): ProtectedLangChainToolCall {
  const call = record(value, "tool call");
  const id = callId(call.id);
  const name = toolCallName(call.name);
  const args = normalizeArguments(call.args, id);
  const encoded = JSON.stringify(args);
  if (byteLength(encoded) > maxArgumentBytes) {
    throw new LangChainProtocolError(`arguments for call ${JSON.stringify(id)} exceed ${maxArgumentBytes} bytes`);
  }
  return { id, name, args, type: typeof call.type === "string" ? call.type : undefined };
}

function normalizeArguments(value: unknown, callIdValue: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new LangChainProtocolError(`arguments for call ${JSON.stringify(callIdValue)} must be an object`);
  }
  return normalizeJSON(value, new WeakSet<object>()) as Record<string, unknown>;
}

function normalizeJSON(value: unknown, ancestors: WeakSet<object>): unknown {
  if (value === null || typeof value === "boolean") return value;
  if (typeof value === "string") { validateUnicode(value, "JSON string"); return value; }
  if (typeof value === "number") {
    if (!Number.isFinite(value)) throw new LangChainProtocolError("JSON number is outside the finite range");
    if (Number.isInteger(value) && !Number.isSafeInteger(value)) {
      throw new LangChainProtocolError("JSON integer is outside the interoperable safe range");
    }
    return value;
  }
  if (typeof value !== "object" || value === undefined) {
    throw new LangChainProtocolError(`unsupported JSON argument type ${typeof value}`);
  }
  if (ancestors.has(value)) throw new LangChainProtocolError("JSON arguments cannot contain cycles");
  ancestors.add(value);
  try {
    if (Array.isArray(value)) {
      const descriptors = Object.getOwnPropertyDescriptors(value);
      const output: unknown[] = [];
      for (let index = 0; index < value.length; index++) {
        const descriptor = descriptors[String(index)];
        if (descriptor === undefined || !("value" in descriptor)) {
          throw new LangChainProtocolError("JSON arrays cannot be sparse or contain accessors");
        }
        output.push(normalizeJSON(descriptor.value, ancestors));
      }
      for (const key of Reflect.ownKeys(descriptors)) {
        if (typeof key === "symbol") throw new LangChainProtocolError("JSON arrays cannot contain symbol properties");
        if (key !== "length" && !/^(0|[1-9]\d*)$/.test(key)) {
          throw new LangChainProtocolError("JSON arrays cannot contain named properties");
        }
      }
      return output;
    }
    const prototype = Object.getPrototypeOf(value);
    if (prototype !== Object.prototype && prototype !== null) {
      throw new LangChainProtocolError("JSON objects must be plain objects");
    }
    const output: Record<string, unknown> = {};
    for (const key of Reflect.ownKeys(value)) {
      if (typeof key !== "string") throw new LangChainProtocolError("JSON objects cannot contain symbol keys");
      validateUnicode(key, "JSON object key");
      const descriptor = Object.getOwnPropertyDescriptor(value, key)!;
      if (!("value" in descriptor) || !descriptor.enumerable) {
        throw new LangChainProtocolError("JSON objects cannot contain accessors or hidden properties");
      }
      Object.defineProperty(output, key, {
        value: normalizeJSON(descriptor.value, ancestors), enumerable: true, writable: true, configurable: true,
      });
    }
    return output;
  } finally {
    ancestors.delete(value);
  }
}

function callId(value: unknown): string {
  if (typeof value !== "string" || value.length < 1 || value.length > 256) {
    throw new LangChainProtocolError("tool call id must be 1-256 visible ASCII characters");
  }
  for (const character of value) {
    const code = character.codePointAt(0)!;
    if (code < 0x21 || code > 0x7e) throw new LangChainProtocolError("tool call id must be 1-256 visible ASCII characters");
  }
  return value;
}

function toolCallName(value: unknown): string {
  if (typeof value !== "string" || value.trim() === "") throw new LangChainProtocolError("tool name must be a non-empty string");
  validateUnicode(value, "tool name");
  if (byteLength(value) > 256) throw new LangChainProtocolError("tool name cannot exceed 256 UTF-8 bytes");
  return value;
}

function toolName(value: unknown): string {
  if ((typeof value !== "object" || value === null) && typeof value !== "function") {
    throw new TypeError("LangGraph tools must expose a name");
  }
  return toolCallName((value as { name?: unknown }).name);
}

function validateUnicode(value: string, name: string): void {
  for (const character of value) {
    const code = character.codePointAt(0)!;
    if (code === 0xfffd || (character.length === 1 && code >= 0xd800 && code <= 0xdfff)) {
      throw new LangChainProtocolError(`${name} cannot contain replacement or unpaired-surrogate characters`);
    }
  }
}

function boundedInteger(value: number, minimum: number, maximum: number, name: string): number {
  if (!Number.isSafeInteger(value) || value < minimum || value > maximum) {
    throw new TypeError(`${name} must be an integer between ${minimum} and ${maximum}`);
  }
  return value;
}

function byteLength(value: string): number {
  return new TextEncoder().encode(value).byteLength;
}

function record(value: unknown, name: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new LangChainProtocolError(`${name} must be an object`);
  }
  return value as Record<string, unknown>;
}

function recordOrUndefined(value: unknown): Record<string, unknown> | undefined {
  return value !== null && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : undefined;
}

async function optionalImport(specifier: string, installation: string): Promise<Record<string, unknown>> {
  try {
    return await import(specifier) as Record<string, unknown>;
  } catch (error) {
    throw new LangChainDependencyError(`${installation}. ${errorMessage(error)}`);
  }
}

function errorMessage(value: unknown): string {
  return value instanceof Error ? value.message : String(value);
}

File: sdk/python/src/latch_sdk/langchain.py
"""Native fail-closed integration for LangChain agents and LangGraph ToolNode."""

import asyncio
from collections import OrderedDict
from collections.abc import Callable, Mapping, Sequence
import json
import math
from threading import Lock
from typing import Any, Dict, List, Optional, Set

try:
    from langchain.agents.middleware import AgentMiddleware
    from langgraph.prebuilt import ToolNode
except ImportError as error:  # pragma: no cover - exercised without the optional extra
    raise ImportError(
        'LangChain support requires: pip install "latch-sdk[langchain]"'
    ) from error

from .errors import LatchError
from .models import Action, Decision


class LangChainProtocolError(LatchError):
    """A framework tool call was unsafe or did not satisfy the adapter contract."""


class LangChainReplayError(LatchError):
    def __init__(self, call_id: str) -> None:
        self.call_id = call_id
        super().__init__(f"LangChain tool call {call_id!r} was duplicated or replayed; execution stopped")


class LangChainNotAllowedError(LatchError):
    def __init__(self, call_id: str, tool: str, decision: Decision) -> None:
        self.call_id = call_id
        self.tool = tool
        self.decision = decision
        super().__init__(
            f"Latch decision for LangChain tool {tool!r} call {call_id!r} is "
            f"{decision.decision}; execution stopped"
        )


class UnknownLangChainToolError(LatchError):
    def __init__(self, call_id: str, tool: str) -> None:
        self.call_id = call_id
        self.tool = tool
        super().__init__(f"LangChain tool {tool!r} for call {call_id!r} is not registered; batch not executed")


class _ToolCall:
    __slots__ = ("call_id", "name", "arguments")

    def __init__(self, call_id: str, name: str, arguments: Dict[str, Any]) -> None:
        self.call_id = call_id
        self.name = name
        self.arguments = arguments


class LangChainToolGuard:
    """Framework-neutral guard used by both middleware and protected ToolNode."""

    def __init__(
        self,
        client: Any,
        *,
        max_calls: int = 128,
        max_argument_bytes: int = 1 << 20,
        replay_capacity: int = 10_000,
    ) -> None:
        if client is None or not callable(getattr(client, "decide", None)):
            raise TypeError("client must provide decide(action)")
        self._max_calls = _bounded_integer(max_calls, 1, 1024, "max_calls")
        self._max_argument_bytes = _bounded_integer(
            max_argument_bytes, 1, 16 << 20, "max_argument_bytes"
        )
        self._replay_capacity = _bounded_integer(replay_capacity, 128, 1_000_000, "replay_capacity")
        if self._replay_capacity < self._max_calls:
            raise ValueError("replay_capacity cannot be smaller than max_calls")
        self._client = client
        self._replayed: "OrderedDict[str, None]" = OrderedDict()
        self._replay_lock = Lock()

    def protect(
        self,
        raw_calls: Sequence[Any],
        *,
        surface: str,
        allowed_tools: Optional[Set[str]] = None,
    ) -> List[_ToolCall]:
        """Validate, authorize, and atomically consume a complete call batch."""
        if isinstance(raw_calls, (str, bytes, bytearray)) or not isinstance(raw_calls, Sequence):
            raise LangChainProtocolError("tool calls must be an array")
        if len(raw_calls) > self._max_calls:
            raise LangChainProtocolError(f"tool call batch exceeds {self._max_calls} calls")
        calls = [self._normalize_call(raw) for raw in raw_calls]
        seen: Set[str] = set()
        for call in calls:
            if call.call_id in seen:
                raise LangChainReplayError(call.call_id)
            seen.add(call.call_id)
            if allowed_tools is not None and call.name not in allowed_tools:
                raise UnknownLangChainToolError(call.call_id, call.name)
        for call in calls:
            decision = self._client.decide(
                Action(
                    tool=call.name,
                    arguments=call.arguments,
                    metadata={
                        "protocol": "langchain",
                        "surface": surface,
                        "tool_call_id": call.call_id,
                        "tool_call_kind": "function",
                    },
                )
            )
            if not decision.allowed:
                raise LangChainNotAllowedError(call.call_id, call.name, decision)
        self._reserve(calls)
        return calls

    async def aprotect(
        self,
        raw_calls: Sequence[Any],
        *,
        surface: str,
        allowed_tools: Optional[Set[str]] = None,
    ) -> List[_ToolCall]:
        return await asyncio.to_thread(
            self.protect, raw_calls, surface=surface, allowed_tools=allowed_tools
        )

    def _normalize_call(self, value: Any) -> _ToolCall:
        call = _mapping(value, "tool call")
        call_id = _call_id(call.get("id"))
        name = _tool_name(call.get("name"))
        arguments = _normalize_arguments(call.get("args"), self._max_argument_bytes, call_id)
        return _ToolCall(call_id, name, arguments)

    def _reserve(self, calls: Sequence[_ToolCall]) -> None:
        with self._replay_lock:
            for call in calls:
                if call.call_id in self._replayed:
                    raise LangChainReplayError(call.call_id)
            for call in calls:
                self._replayed[call.call_id] = None
            while len(self._replayed) > self._replay_capacity:
                self._replayed.popitem(last=False)


class LatchAgentMiddleware(AgentMiddleware):
    """Drop-in middleware for ``create_agent(..., middleware=[...])``."""

    def __init__(self, client: Any, **guard_options: Any) -> None:
        super().__init__()
        self._latch_guard = LangChainToolGuard(client, **guard_options)

    def wrap_tool_call(self, request: Any, handler: Callable[[Any], Any]) -> Any:
        if getattr(request, "tool", None) is None:
            raise LangChainProtocolError("tool call has no resolved LangChain tool")
        self._latch_guard.protect(
            [getattr(request, "tool_call", None)], surface="langchain_agent"
        )
        return handler(request)

    async def awrap_tool_call(self, request: Any, handler: Callable[[Any], Any]) -> Any:
        if getattr(request, "tool", None) is None:
            raise LangChainProtocolError("tool call has no resolved LangChain tool")
        await self._latch_guard.aprotect(
            [getattr(request, "tool_call", None)], surface="langchain_agent"
        )
        return await handler(request)


class LatchToolNode(ToolNode):
    """LangGraph ToolNode that authorizes the full parallel batch first."""

    def __init__(
        self,
        tools: Sequence[Any],
        client: Any,
        *,
        latch_max_calls: int = 128,
        latch_max_argument_bytes: int = 1 << 20,
        latch_replay_capacity: int = 10_000,
        **kwargs: Any,
    ) -> None:
        super().__init__(tools, **kwargs)
        self._latch_guard = LangChainToolGuard(
            client,
            max_calls=latch_max_calls,
            max_argument_bytes=latch_max_argument_bytes,
            replay_capacity=latch_replay_capacity,
        )
        self._latch_messages_key = kwargs.get("messages_key", "messages")
        self._latch_allowed_tools = set(self.tools_by_name)

    def invoke(self, input: Any, config: Any = None, **kwargs: Any) -> Any:
        calls = _extract_tool_calls(input, self._latch_messages_key)
        self._latch_guard.protect(
            calls,
            surface="langgraph_tool_node",
            allowed_tools=self._latch_allowed_tools,
        )
        return super().invoke(input, config, **kwargs)

    async def ainvoke(self, input: Any, config: Any = None, **kwargs: Any) -> Any:
        calls = _extract_tool_calls(input, self._latch_messages_key)
        await self._latch_guard.aprotect(
            calls,
            surface="langgraph_tool_node",
            allowed_tools=self._latch_allowed_tools,
        )
        return await super().ainvoke(input, config, **kwargs)


def _extract_tool_calls(value: Any, messages_key: str) -> Sequence[Any]:
    if isinstance(value, list):
        if not value:
            return []
        if all(isinstance(item, Mapping) and {"id", "name", "args"}.issubset(item) for item in value):
            return value
        message = value[-1]
    elif isinstance(value, Mapping):
        messages = value.get(messages_key)
        if not isinstance(messages, list) or not messages:
            raise LangChainProtocolError(f"LangGraph state must contain a non-empty {messages_key!r} list")
        message = messages[-1]
    else:
        messages = getattr(value, messages_key, None)
        if not isinstance(messages, list) or not messages:
            raise LangChainProtocolError(f"LangGraph state must contain a non-empty {messages_key!r} list")
        message = messages[-1]
    calls = message.get("tool_calls") if isinstance(message, Mapping) else getattr(message, "tool_calls", None)
    if calls is None:
        return []
    if isinstance(calls, (str, bytes, bytearray)) or not isinstance(calls, Sequence):
        raise LangChainProtocolError("the last message tool_calls value must be an array")
    return calls


def _normalize_arguments(value: Any, limit: int, call_id: str) -> Dict[str, Any]:
    if not isinstance(value, dict):
        raise LangChainProtocolError(f"arguments for call {call_id!r} must be an object")
    normalized = _normalize_json(value, set())
    try:
        payload = json.dumps(normalized, separators=(",", ":"), ensure_ascii=False, allow_nan=False).encode("utf-8")
    except (TypeError, ValueError, UnicodeError) as error:
        raise LangChainProtocolError(f"arguments for call {call_id!r} are not interoperable JSON: {error}") from error
    if len(payload) > limit:
        raise LangChainProtocolError(f"arguments for call {call_id!r} exceed {limit} bytes")
    return normalized


def _normalize_json(value: Any, ancestors: Set[int]) -> Any:
    if value is None or isinstance(value, bool):
        return value
    if isinstance(value, str):
        _validate_unicode(value, "JSON string")
        return value
    if isinstance(value, int):
        if abs(value) > 9_007_199_254_740_991:
            raise LangChainProtocolError("JSON integer is outside the interoperable safe range")
        return value
    if isinstance(value, float):
        if not math.isfinite(value):
            raise LangChainProtocolError("JSON number is outside the finite range")
        if value.is_integer() and abs(value) > 9_007_199_254_740_991:
            raise LangChainProtocolError("JSON integer-valued number is outside the interoperable safe range")
        return value
    if isinstance(value, (list, dict)):
        identity = id(value)
        if identity in ancestors:
            raise LangChainProtocolError("JSON arguments cannot contain cycles")
        ancestors.add(identity)
        try:
            if isinstance(value, list):
                return [_normalize_json(item, ancestors) for item in value]
            result: Dict[str, Any] = {}
            for key, item in value.items():
                if not isinstance(key, str):
                    raise LangChainProtocolError("JSON object keys must be strings")
                _validate_unicode(key, "JSON object key")
                result[key] = _normalize_json(item, ancestors)
            return result
        finally:
            ancestors.remove(identity)
    raise LangChainProtocolError(f"unsupported JSON argument type {type(value).__name__}")


def _mapping(value: Any, name: str) -> Mapping[str, Any]:
    if not isinstance(value, Mapping):
        raise LangChainProtocolError(f"{name} must be an object")
    return value


def _call_id(value: Any) -> str:
    if not isinstance(value, str) or not 1 <= len(value) <= 256:
        raise LangChainProtocolError("tool call id must be 1-256 visible ASCII characters")
    if any(ord(character) < 0x21 or ord(character) > 0x7E for character in value):
        raise LangChainProtocolError("tool call id must be 1-256 visible ASCII characters")
    return value


def _tool_name(value: Any) -> str:
    if not isinstance(value, str) or not value.strip():
        raise LangChainProtocolError("tool name must be a non-empty string")
    _validate_unicode(value, "tool name")
    if len(value.encode("utf-8")) > 256:
        raise LangChainProtocolError("tool name cannot exceed 256 UTF-8 bytes")
    return value


def _validate_unicode(value: str, name: str) -> None:
    if any(ord(character) == 0xFFFD or 0xD800 <= ord(character) <= 0xDFFF for character in value):
        raise LangChainProtocolError(f"{name} cannot contain replacement or unpaired-surrogate characters")


def _bounded_integer(value: Any, minimum: int, maximum: int, name: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or not minimum <= value <= maximum:
        raise ValueError(f"{name} must be between {minimum} and {maximum}")
    return value


__all__ = [
    "LangChainNotAllowedError",
    "LangChainProtocolError",
    "LangChainReplayError",
    "LangChainToolGuard",
    "LatchAgentMiddleware",
    "LatchToolNode",
    "UnknownLangChainToolError",
]
